### PR TITLE
Centos singularity

### PR DESCRIPTION
**Pull request by bhummerstone**
Allows the creation of a CentOS HPC VM with Singularity installed. Different versions of CentOS/Singularity can be installed by passing parameters (defaults are 7.3 and 2.3.1 respectively)

### CODE REVIEW

File: centos-singularity/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    "addressPrefix": "10.0.0.0/16",
+    "subnetName": "Subnet",
+    "subnetPrefix": "10.0.0.0/24",
+    "storageAccountType": "Standard_LRS",
+    "publicIPAddressName": "myPublicIP",
+    "publicIPAddressType": "Dynamic",
+    "vmName": "MySingularityVM",
+    "virtualNetworkName": "MyVNET",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',variables('virtualNetworkName'))]",
+    "subnetRef": "[concat(variables('vnetID'),'/subnets/',variables('subnetName'))]"
```
Comment (bmoore-msft) [inline centos-singularity/azuredeploy.json]: simplify, just use resourceId()

File: centos-singularity/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      "tags": {
+          "displayName": "InstallSingularity"
+      },
+      "properties": {
+          "publisher": "Microsoft.Azure.Extensions",
+          "type": "CustomScript",
+          "typeHandlerVersion": "2.0",
+          "autoUpgradeMinorVersion": true,
+        "settings": {
+          "fileUris": [
```
Comment (bmoore-msft) [inline centos-singularity/azuredeploy.json]: see BP guide for how to handle artifacts

File: centos-singularity/azuredeploy.parameters.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "adminUsername": {
+      "value": "azureuser"
```
Comment (bmoore-msft) [inline centos-singularity/azuredeploy.parameters.json]: GEN-UNIQUE to reduce surface area

File: centos-singularity/metadata.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "itemDisplayName": "Deploy a simple Linux VM",
+  "icon": "ubuntu",
+  "description": "This template allows you to deploy a simple Linux VM using a few different options for the Ubuntu version, using the latest patched version. This will deploy a A1 size VM in the resource group location and return the FQDN of the VM.",
+  "summary": "This template takes a minimum amount of parameters and deploys a Linux VM, using the latest patched version.",
+  "githubUsername": "bmoore-msft",
```
Comment (bmoore-msft) [inline centos-singularity/metadata.json]: update username

**Comment by bhummerstone** [comment]
OK, done: thanks Brian

File: centos-singularity/azuredeploy.json (hunk 1 above)
Comment (bhummerstone) [inline centos-singularity/azuredeploy.json]: I tried this, but the CI build failed; I can't see a way to do a resourceId for the subnet, as it is dependent on the name of the vNet

Comment (bmoore-msft) [inline centos-singularity/azuredeploy.json]: try: resourceId('Microsoft.Network/virtualNetworks/subnets', variables('virtualNetworkName'), variables('subnetName')) - and then you can remove the vnetId variable.

File: centos-singularity/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            "[parameters('_artifactsLocation')]"
```
Comment (bmoore-msft) [inline centos-singularity/azuredeploy.json]: this should be multi-part and secured with a sasToken - so you'll end up with something like:

      "fileUris": [
          "[concat(parameters('_artifactsLocation'), '/', variables('ScriptFolder'), '/', variables('ScriptFileName'), parameters('_artifactsLocationSasToken'))]"
        ],

See: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#samples-that-contain-extra-artifacts-custom-scripts-nested-templates-etc

**Comment by bhummerstone** [comment]
CI is now failing with "Server Error:{
    "error": {
        "code": "ECONNREFUSED",
        "errno": "ECONNREFUSED",
        "syscall": "connect"
    }
}"

**Pull request merged**